Live Pixel Validator

ladamski opened the pull request:
https://app.asana.com/0/1205243787707480/1207703134691882/f

nshuba commented on src/params_validator.mjs:
Can probably remove these 2 comments

nshuba commented on tests/live_pixel_validation_test.mjs:
```diff
@@ -115,6 +115,7 @@ describe('Common suffixes', () => {
                     enum: [1, 2, 3],
                 },
             ],
+            parameters: [],
```
We should update the [schema](https://github.com/duckduckgo/pixel-schema/blob/main/schemas/pixel_schema.json5#L13) itself then if we'll be requiring 'parameters' to be defined.

ladamski replied:
Great catch! I just updated the PR to support empty params instead, I think easier?

nshuba commented on src/error_utils.mjs:
```diff
@@ -19,7 +19,16 @@ function formatAjvErrors(validationErrors) {
+                formattedError += `. Found extra suffix '${suffixes[error.params.additionalProperty]}' at index ${error.params.additionalProperty}`;
+            } else if (error.params.allowedValues) {
+                const idx = error.instancePath ? Number(error.instancePath.split('/')[1]) : error.params.additionalProperty;
+                formattedError = `Suffix '${suffixes[idx]}' at index ${idx} ` + formattedError;
```
Very minor: I think this prints the index twice? It's already in `formattedError`, so we can likely get rid of `${idx}` in this case

ladamski replied:
I added this because not all permutations of `formattedError` include it, so I'd rather have some duplication then have to guess.

nshuba replied:
Ah that makes sense! Thanks for explaining :)

nshuba commented on src/error_utils.mjs:
```diff
@@ -8,7 +8,7 @@
  * @param {Array<import("ajv").ErrorObject>} validationErrors - array of AJV error objects
  * @returns {Array<string>} - array of formatted error messages
```
Update doc string? :) 🙏

nshuba commented on tests/test_data/invalid/product.json:
```diff
@@ -0,0 +1,8 @@
+        "key": "appVersion",
+        "version": "0.98.4"
+    },
+    "forceLowerCase": false
```
Should we have a test case for case insensitive as well? Can be a quick test, with a `product.json` defined inside the test case.

nshuba reviewed (approved):
Left some comments - nothing major really, so approving 👍

ladamski commented:
@nshuba I've addressed your comments above, appreciate the detailed review!  I could merge it but wanted to give you a chance to take another quick pass if you want to, otherwise I can just merge it.

nshuba commented on tests/test_data/valid_case_insensitive/pixels/pixel_guide.json5:
```diff
@@ -0,0 +1,69 @@
+{
```
Probably don't need to duplicate the guide here

nshuba commented on tests/test_data/valid_case_insensitive/pixels/pixel_subfolder/test_pixels.json:
```diff
@@ -0,0 +1,64 @@
+{
+    // Collection of pixels from various platforms for testing
+    // Note: definitions may not be fully accurate
+    "m.windows.crash.stable": {
```
Just to keep it simple, we can probably keep a simplified version of the `crash` pixel only and also keep this file outside the sub-folder (we test nested folders elsewhere)

nshuba reviewed (approved):
Thanks for addressing all the comments!
